core/src/main/scala/org/apache/spark/util/collection/ReferenceCounter.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util.collection
+
+import scala.collection.JavaConverters._
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.common.collect.ConcurrentHashMultiset
+
+import org.apache.spark.TaskContext
+
+/**
+ * Thread-safe collection for maintaining both global and per-task reference counts for objects.
+ */
+private[spark] class ReferenceCounter[T] {
```

Comment:
Is there any reason you did it this way instead of a counter per object? Not sure how many blocks we have but this seems contention prone.